File: js/tests/match.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import { compile } from "../src/compile.js";
import { createRegexStrategy } from "../src/strategies/regex.js";
import { createTrieStrategy } from "../src/strategies/trie.js";

/**
 * Build both matching strategies for the same route collection so every
 * assertion can be verified against the regex and trie implementations.
 *
 * @param {Object<string, any>} routes - Map of pattern strings to opaque values.
 * @returns {import("../src/compile.js").MatchStrategy[]} The regex and trie strategies.
 */
function buildStrategies(routes) {
  const compiled = compile(routes);
  return [createRegexStrategy(compiled), createTrieStrategy(compiled)];
}

for (const [name, strategyFactory] of [
  ["regex", createRegexStrategy],
  ["trie", createTrieStrategy],
]) {
  test(`[${name}] matches static root`, () => {
    const [strategy] = buildStrategies({ "/": "home" });
    const result = strategy.match("/");
    assert.equal(result.found, true);
    assert.equal(result.value, "home");
    assert.deepEqual(result.params, {});
  });

  test(`[${name}] matches static route`, () => {
    const [strategy] = buildStrategies({ "/about": "about" });
    const result = strategy.match("/about");
    assert.equal(result.found, true);
    assert.equal(result.route, "/about");
  });

  test(`[${name}] matches dynamic route and extracts params`, () => {
    const [strategy] = buildStrategies({ "/articles/:slug": "page" });
    const result = strategy.match("/articles/javascript");
    assert.equal(result.found, true);
    assert.equal(result.route, "/articles/:slug");
    assert.deepEqual(result.params, { slug: "javascript" });
  });

  test(`[${name}] returns not found when no route matches`, () => {
    const [strategy] = buildStrategies({ "/about": "about" });
    const result = strategy.match("/missing");
    assert.equal(result.found, false);
    assert.equal(result.route, null);
    assert.equal(result.value, null);
  });

  test(`[${name}] falls back to wildcard`, () => {
    const [strategy] = buildStrategies({ "*": "404" });
    const result = strategy.match("/articles/javascript");
    assert.equal(result.found, true);
    assert.equal(result.route, "*");
    assert.equal(result.value, "404");
  });

  test(`[${name}] static takes priority over wildcard`, () => {
    const [strategy] = buildStrategies({ "/about": "about", "*": "404" });
    const result = strategy.match("/about");
    assert.equal(result.value, "about");
  });

  test(`[${name}] dynamic takes priority over wildcard`, () => {
    const [strategy] = buildStrategies({ "/articles/:slug": "page", "*": "404" });
    const result = strategy.match("/articles/javascript");
    assert.equal(result.value, "page");
  });

  test(`[${name}] normalizes query string before matching`, () => {
    const [strategy] = buildStrategies({ "/articles/:slug": "page" });
    const result = strategy.match("/articles/javascript?page=1");
    assert.equal(result.found, true);
    assert.equal(result.pathname, "/articles/javascript");
  });

  test(`[${name}] parses query into the result`, () => {
    const [strategy] = buildStrategies({ "/articles/:slug": "page" });
    const result = strategy.match("/articles/javascript?page=1&perPage=10");
    assert.deepEqual(result.query, { page: "1", perPage: "10" });
    assert.deepEqual(result.params, { slug: "javascript" });
  });

  test(`[${name}] returns empty query when none present`, () => {
    const [strategy] = buildStrategies({ "/about": "about" });
    const result = strategy.match("/about");
    assert.deepEqual(result.query, {});
  });

  test(`[${name}] collects repeated query keys into an array`, () => {
    const [strategy] = buildStrategies({ "/search": "s" });
    const result = strategy.match("/search?tag=a&tag=b");
    assert.deepEqual(result.query, { tag: ["a", "b"] });
  });

  test(`[${name}] normalizes fragment before matching`, () => {
    const [strategy] = buildStrategies({ "/articles": "page" });
    const result = strategy.match("/articles#section");
    assert.equal(result.found, true);
    assert.equal(result.pathname, "/articles");
  });

  test(`[${name}] decodes url-encoded params`, () => {
    const [strategy] = buildStrategies({ "/files/:name": "file" });
    const result = strategy.match("/files/hello%20world");
    assert.deepEqual(result.params, { name: "hello world" });
  });

  test(`[${name}] segment count must match for dynamic routes`, () => {
    const [strategy] = buildStrategies({ "/users/:id": "u" });
    const result = strategy.match("/users/1/posts");
    assert.equal(result.found, false);
  });

  test(`[${name}] returns composed array value untouched`, () => {
    const value = ["a", "b", "c"];
    const [strategy] = buildStrategies({ "/composed": value });
    const result = strategy.match("/composed");
    assert.equal(result.value, value);
  });

  test(`[${name}] static takes priority over dynamic at same level`, () => {
    const [strategy] = buildStrategies({
      "/users/admin": "admin",
      "/users/:id": "user",
    });
    const result = strategy.match("/users/admin");
    assert.equal(result.route, "/users/admin");
    assert.deepEqual(result.params, {});
  });
}

File: go/match_test.go
package roution

import (
	"reflect"
	"testing"
)

// buildStrategies returns both matching strategies for the same routes so
// every assertion can be verified against the regex and trie implementations.
func buildStrategies(t *testing.T, routes map[string]any) []Matcher {
	t.Helper()
	compiled := Compile(routes)
	return []Matcher{
		NewRegexMatcher(compiled, true),
		NewTrieMatcher(compiled, true),
	}
}

func TestStrategiesMatchStaticRoot(t *testing.T) {
	for _, m := range buildStrategies(t, map[string]any{"/": "home"}) {
		r := m.Match("/")
		if !r.Found || r.Value != "home" {
			t.Fatalf("%T: %+v", m, r)
		}
		if len(r.Params) != 0 {
			t.Fatalf("%T: params = %v, want empty", m, r.Params)
		}
	}
}

func TestStrategiesMatchStaticRoute(t *testing.T) {
	for _, m := range buildStrategies(t, map[string]any{"/about": "about"}) {
		r := m.Match("/about")
		if !r.Found || r.Route != "/about" {
			t.Fatalf("%T: %+v", m, r)
		}
	}
}

func TestStrategiesMatchDynamicAndExtractParams(t *testing.T) {
	for _, m := range buildStrategies(t, map[string]any{"/articles/:slug": "page"}) {
		r := m.Match("/articles/javascript")
		if !r.Found || r.Route != "/articles/:slug" {
			t.Fatalf("%T: %+v", m, r)
		}
		if !reflect.DeepEqual(r.Params, map[string]string{"slug": "javascript"}) {
			t.Fatalf("%T: params = %v", m, r.Params)
		}
	}
}

func TestStrategiesNotFound(t *testing.T) {
	for _, m := range buildStrategies(t, map[string]any{"/about": "about"}) {
		r := m.Match("/missing")
		if r.Found || r.Route != "" || r.Value != nil {
			t.Fatalf("%T: %+v", m, r)
		}
	}
}

func TestStrategiesWildcardFallback(t *testing.T) {
	for _, m := range buildStrategies(t, map[string]any{"*": "404"}) {
		r := m.Match("/articles/javascript")
		if !r.Found || r.Route != "*" || r.Value != "404" {
			t.Fatalf("%T: %+v", m, r)
		}
	}
}

func TestStrategiesStaticPriorityOverWildcard(t *testing.T) {
	for _, m := range buildStrategies(t, map[string]any{"/about": "about", "*": "404"}) {
		if r := m.Match("/about"); r.Value != "about" {
			t.Fatalf("%T: %+v", m, r)
		}
	}
}

func TestStrategiesDynamicPriorityOverWildcard(t *testing.T) {
	for _, m := range buildStrategies(t, map[string]any{"/articles/:slug": "page", "*": "404"}) {
		if r := m.Match("/articles/javascript"); r.Value != "page" {
			t.Fatalf("%T: %+v", m, r)
		}
	}
}

func TestStrategiesNormalizeQuery(t *testing.T) {
	for _, m := range buildStrategies(t, map[string]any{"/articles/:slug": "page"}) {
		r := m.Match("/articles/javascript?page=1")
		if !r.Found || r.Pathname != "/articles/javascript" {
			t.Fatalf("%T: %+v", m, r)
		}
	}
}

func TestStrategiesNormalizeFragment(t *testing.T) {
	for _, m := range buildStrategies(t, map[string]any{"/articles": "page"}) {
		r := m.Match("/articles#section")
		if !r.Found || r.Pathname != "/articles" {
			t.Fatalf("%T: %+v", m, r)
		}
	}
}

func TestStrategiesDecodeParams(t *testing.T) {
	for _, m := range buildStrategies(t, map[string]any{"/files/:name": "file"}) {
		r := m.Match("/files/hello%20world")
		if !reflect.DeepEqual(r.Params, map[string]string{"name": "hello world"}) {
			t.Fatalf("%T: params = %v", m, r.Params)
		}
	}
}

func TestStrategiesSegmentCountMustMatch(t *testing.T) {
	for _, m := range buildStrategies(t, map[string]any{"/users/:id": "u"}) {
		if r := m.Match("/users/1/posts"); r.Found {
			t.Fatalf("%T: expected no match, got %+v", m, r)
		}
	}
}

func TestStrategiesStaticPriorityOverDynamic(t *testing.T) {
	for _, m := range buildStrategies(t, map[string]any{
		"/users/admin": "admin",
		"/users/:id":   "user",
	}) {
		r := m.Match("/users/admin")
		if r.Route != "/users/admin" {
			t.Fatalf("%T: route = %s, want /users/admin", m, r.Route)
		}
		if len(r.Params) != 0 {
			t.Fatalf("%T: params = %v, want empty", m, r.Params)
		}
	}
}

func TestStrategiesIdenticalResults(t *testing.T) {
	routes := map[string]any{
		"/":                  "home",
		"/about":             "about",
		"/articles":          "list",
		"/articles/:slug":    "article",
		"/users/:id/posts/:postId": "post",
		"/composed":          []string{"a", "b"},
		"/login":             map[string]any{"auth": false},
		"/version":           "1.0.0",
		"/numbers":           []int{1, 2, 3},
		"*":                  "404",
	}
	compiled := Compile(routes)
	regex := NewRegexMatcher(compiled, true)
	trie := NewTrieMatcher(compiled, true)

	samples := []string{
		"/",
		"/about",
		"/articles",
		"/articles/javascript?page=1&perPage=10",
		"/users/42/posts/7#comments",
		"/composed",
		"/login",
		"/version",
		"/numbers",
		"/articles/javascript/extra",
		"/missing/page",
	}

	for _, p := range samples {
		a := regex.Match(p)
		b := trie.Match(p)
		if !reflect.DeepEqual(a, b) {
			t.Fatalf("mismatch for %q:\n regex=%+v\n trie =%+v", p, a, b)
		}
	}
}

func TestStrategiesQueryDisabled(t *testing.T) {
	compiled := Compile(map[string]any{"/a": "v"})
	regex := NewRegexMatcher(compiled, false)
	trie := NewTrieMatcher(compiled, false)
	for _, m := range []Matcher{regex, trie} {
		if r := m.Match("/a?x=1"); len(r.Query) != 0 {
			t.Fatalf("%T: query = %v, want empty", m, r.Query)
		}
	}
}
